### PR TITLE
check Poco version and if below 1.4.1p1 build from source

### LINKED ISSUE #9: check version of poco and build from source if older than 1.4.1p1
Based on the bug report https://github.com/ros2/demos/issues/177

`poco_vendor` should be updated to add logic to check if the version detected on the system is recent enough, if not it should go through the code path that  clones and builds poco from source

connects to ros2/demos#177

### LINKED ISSUE #177: pendulum_demo sometimes hanging forever
The test `test_pendulum_teleop` sometimes hangs forever and therefore the test times out. After reproducing this locally with `RMW_IMPLEMENTATION=rmw_connext_cpp python3 build_isolated/pendulum_control/test_pendulum_teleop__rmw_connext_cpp.py` I was able to look at the stack:

```
#0  __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
#1  0x00007f8b6d906dbd in __GI___pthread_mutex_lock (mutex=0x7f8b6c46fc00 <Poco::SharedLibraryImpl::_mutex>) at ../nptl/pthread_mutex_lock.c:80
#2  0x00007f8b6c153a98 in Poco::MutexImpl::lockImpl (this=0x7f8b6c46fc00 <Poco::SharedLibraryImpl::_mutex>) at include/Poco/Mutex_POSIX.h:81
#3  0x00007f8b6c153c86 in Poco::FastMutex::lock (this=0x7f8b6c46fc00 <Poco::SharedLibraryImpl::_mutex>) at include/Poco/Mutex.h:205
#4  0x00007f8b6c1544c0 in Poco::ScopedLock<Poco::FastMutex>::ScopedLock (this=0x7ffc2fe659b0, mutex=...) at include/Poco/ScopedLock.h:59
#5  0x00007f8b6c1bf5e8 in Poco::SharedLibraryImpl::findSymbolImpl (this=0xf39ba8, name="rmw_trigger_guard_condition") at src/SharedLibrary_UNIX.cpp:100
#6  0x00007f8b6c1bf8a1 in Poco::SharedLibrary::hasSymbol (this=0xf39ba0, name="rmw_trigger_guard_condition") at src/SharedLibrary.cpp:93
#7  0x00007f8b6c680839 in get_symbol (symbol_name=0x7f8b6c681f70 "rmw_trigger_guard_condition") at ../ws_ros2/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:148
#8  0x00007f8b6c681076 in rmw_trigger_guard_condition (guard_condition=0xf80e00) at ../ws_ros2/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:423
#9  0x00007f8b6e360963 in rcl_trigger_guard_condition (guard_condition=0x171e4a8) at ../ws_ros2/src/ros2/rcl/rcl/src/rcl/guard_condition.c:154
#10 0x00007f8b6e0c9c8d in trigger_interrupt_guard_condition (signal_value=2) at ../ws_ros2/src/ros2/rclcpp/rclcpp/src/rclcpp/utilities.cpp:115
#11 0x00007f8b6e0c9dc7 in signal_handler (signal_value=2, siginfo=0x7ffc2fe65cf0, context=0x7ffc2fe65bc0) at ../ws_ros2/src/ros2/rclcpp/rclcpp/src/rclcpp/utilities.cpp:158
#12 <signal handler called>
#13 0x00007f8b6c1bf5f8 in Poco::SharedLibraryImpl::findSymbolImpl (this=0xf39ba8, name="rmw_wait") at src/SharedLibrary_UNIX.cpp:103
#14 0x00007f8b6c1bf8a1 in Poco::SharedLibrary::hasSymbol (this=0xf39ba0, name="rmw_wait") at src/SharedLibrary.cpp:93
#15 0x00007f8b6c680839 in get_symbol (symbol_name=0x7f8b6c681fb3 "rmw_wait") at ../ws_ros2/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:148
#16 0x00007f8b6c68114d in rmw_wait (subscriptions=0x7f8b239fb908, guard_conditions=0x7f8b239fb920, services=0x7f8b239fb950, clients=0x7f8b239fb938, waitset=0x1f122f0, wait_timeout=0x7ffc2fe66450)
    at ../ws_ros2/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:456
#17 0x00007f8b6e36b06a in rcl_wait (wait_set=0x1f12878, timeout=0) at ../ws_ros2/src/ros2/rcl/rcl/src/rcl/wait.c:590
#18 0x00007f8b6e021f98 in rclcpp::executor::Executor::wait_for_work (this=0x1f12860, timeout=...) at ../ws_ros2/src/ros2/rclcpp/rclcpp/src/rclcpp/executor.cpp:409
#19 0x00007f8b6e0231db in rclcpp::executor::Executor::get_next_executable (this=0x1f12860, timeout=...) at ../ws_ros2/src/ros2/rclcpp/rclcpp/src/rclcpp/executor.cpp:544
#20 0x00007f8b6e020b7e in rclcpp::executor::Executor::spin_some (this=0x1f12860) at ../ws_ros2/src/ros2/rclcpp/rclcpp/src/rclcpp/executor.cpp:191
#21 0x000000000042124f in pendulum_control::RttExecutor::loop_callback (arg=0x1f12860) at ../ws_ros2/src/ros2/demos/pendulum_control/include/pendulum_control/rtt_executor.hpp:135
#22 0x00007f8b6db2f821 in Rttest::spin_once (this=0xf35a48, user_function=0x4211f5 <pendulum_control::RttExecutor::loop_callback(void*)>, args=0x1f12860, start_time=0x7ffc2fe66750, update_period=0xf35a50, i=1145)
    at ../ws_ros2/src/ros2/realtime_support/rttest/src/rttest.cpp:595
#23 0x00007f8b6db2f576 in Rttest::spin_period (this=0xf35a48, user_function=0x4211f5 <pendulum_control::RttExecutor::loop_callback(void*)>, args=0x1f12860, update_period=0xf35a50, iterations=0)
    at ../ws_ros2/src/ros2/realtime_support/rttest/src/rttest.cpp:551
#24 0x00007f8b6db2f8b8 in rttest_spin_period (user_function=0x4211f5 <pendulum_control::RttExecutor::loop_callback(void*)>, args=0x1f12860, update_period=0xf35a50, iterations=0)
    at ../ws_ros2/src/ros2/realtime_support/rttest/src/rttest.cpp:608
#25 0x00007f8b6db2f4e9 in Rttest::spin (this=0xf35a48, user_function=0x4211f5 <pendulum_control::RttExecutor::loop_callback(void*)>, args=0x1f12860)
    at ../ws_ros2/src/ros2/realtime_support/rttest/src/rttest.cpp:539
#26 0x00007f8b6db2f3f3 in rttest_spin (user_function=0x4211f5 <pendulum_control::RttExecutor::loop_callback(void*)>, args=0x1f12860) at ../ws_ros2/src/ros2/realtime_support/rttest/src/rttest.cpp:512
#27 0x00000000004211bb in pendulum_control::RttExecutor::spin (this=0x1f12860) at ../ws_ros2/src/ros2/demos/pendulum_control/include/pendulum_control/rtt_executor.hpp:95
#28 0x0000000000415d20 in main (argc=3, argv=0x7ffc2fe66c68) at ../ws_ros2/src/ros2/demos/pendulum_control/src/pendulum_demo.cpp:291
```

The relevant pieces of information are in the following frames:
* (15) While checking for the existence of a symbol which locks a mutex
* (12) the incoming `SIGINT` signal is being handled
* (7) which in order to trigger the guard condition needs to check the existence of a different symbol
* :lock:

If the used mutex would be recursive that wouldn't be a problem. But until Poco [1.4.1p1](https://github.com/pocoproject/poco/blob/b9acb9f263d0f3776d87d0de49ded660ab675061/CHANGELOG#L314-L324) Poco uses a non-recursive mutex on Linux. Since Xenial is using version 1.3.6p1 we see the deadlock.

Possible workarounds:
* Use a newer version of Poco (newer Ubuntu distro have a new enough version).
* Build using a single rmw implementation which avoids the usage of Poco in the `rmw_implemenation` package.

### PR DESCRIPTION
**Pull request by dirk-thomas**
See ros2/demos#177 why we need at least Poco version 1.4.1p1.

This patch falls back to build Poco from source if the installed version is too old which fixes ros2/demos#177: [Build Status](http://ci.ros2.org/job/ci_linux/3258/) (look for `Found Poco version 0x01030600 is too old, building from source instead` in the console output).

Fixes #9

### CODE REVIEW

File: cmake/Modules/FindPoco.cmake [hunk 1]
```diff
@@ -118,6 +118,35 @@ endif()
...
+if(Poco_INCLUDE_DIR)
+  find_file (
+      Poco_VERSION_FILE NAMES "Version.h"
+      PATHS "${Poco_INCLUDE_DIR}"
+      PATH_SUFFIXES "Poco" "Foundation/include/Poco"
+      DOC "Path of Poco/Version.h file"
+    NO_DEFAULT_PATH
+  )
+  if(NOT Poco_VERSION_FILE)
+    message(STATUS "Found Poco version (< 1.4.0) is too old, building from source instead")
```
Comment (mikaelarguedas) [inline cmake/Modules/FindPoco.cmake]: Looks like this version file was [added in 1.3.7](https://github.com/pocoproject/poco/commit/7a563faa29be8dd5a47e4ea2f5aa024e0e27bbcc#diff-a240205fff6d12b8a861b5cc16475a65), can you clarify how you picked "1.4.0" for the printed message?

Comment (dirk-thomas) [reply cmake/Modules/FindPoco.cmake]: The `CHANGELOG` lists the addition in the "Release 1.4.0" section. That is where I got the version number from. Also I don't see any listing of a version 1.3.7 (neither in the changelog nor as a tag).

Comment (mikaelarguedas) [reply cmake/Modules/FindPoco.cmake]: ok, I guess they had a branch for it but didnt ever released 1.3.7 and several commits on that branch say "preparing for 1.4.0". Sorry for the noise

**Review by mikaelarguedas — APPROVED** [review]
Nice! ran 100 times without failing :+1:

**Comment by dirk-thomas** [comment]
Thanks for catching and fixing the style glitch.

**Pull request merged**